File: src/main/java/algo/sequence/HandOfStraights_846.java
package algo.sequence;

import java.util.Arrays;
import java.util.HashMap;

public class HandOfStraights_846 {
  public static boolean isNStraightHand(int[] hand, int groupSize) {
    var len = hand.length;
    if (len < groupSize || len % groupSize != 0) {
      return false;
    }
    if (len == 1 || groupSize == 1) {
      return true;
    }

    Arrays.sort(hand);

    var freq = new HashMap<Integer, Integer>();
    for (var n : hand) {
      freq.put(n, freq.getOrDefault(n, 0) + 1);
    }

    var groupNum = len / groupSize;
    var groupCount = 0;
    for (var h : hand) {
      if (freq.get(h) > 0) {
        if (groupCount > groupNum) {
          return false;
        }
        for (var i = 0; i < groupSize; i++) {
          var next = h + i;
          if (!freq.containsKey(next) || freq.get(next) == 0) {
            return false;
          }
          freq.put(next, freq.get(next) - 1);
        }
        groupCount++;
      }
    }

    return groupCount == groupNum;
  }

  public static void main(String[] args) {
    System.out.println(isNStraightHand(new int[]{1,2,3,6,2,3,4,7,8}, 3));
    System.out.println(isNStraightHand(new int[]{1,2}, 2));
    System.out.println(isNStraightHand(new int[]{6,7,5,3,4,7,8,10,9,6}, 5));
  }
}

File: src/main/kotlin/algo/sequence/HandOfStraights_846.kt
package algo.sequence

class KtHandOfStraights_846 {
    fun isNStraightHand(hand: IntArray, groupSize: Int): Boolean {
        val freq: MutableMap<Int, Int> = mutableMapOf()
        hand.forEach {
            freq[it] = freq.getOrDefault(it, 0) + 1
        }

        hand.sort()
        val groupNum = hand.size / groupSize
        var groupCount = 0
        hand.forEach { num ->
            if (groupCount > groupNum) {
                return false
            }
            if (freq[num]!! == 0) {
                return@forEach
            }
            var card = num
            for (i in 1..groupSize) {
                if (freq[card] != null && freq[card]!! > 0) {
                    freq[card] = freq[card]!! - 1
                } else {
                    return false
                }
                card++
            }
            groupCount++
        }
        return groupCount == groupNum
    }
}

fun main() {
    val hand = KtHandOfStraights_846()
    println(hand.isNStraightHand(intArrayOf(1,2,3,6,2,3,4,7,8), 3))
    println(hand.isNStraightHand(intArrayOf(1,2), 2))
}